File: tests/client_compatibility/go/query_test.go
package main

import (
	"context"
	"testing"

	"github.com/jackc/pgx/v5"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

/**
 * Test simple query execution via pgx driver.
 *
 * Tests P1 Simple Query Protocol:
 * - Query message handling
 * - IRIS SQL execution
 * - Row data encoding
 * - CommandComplete and ReadyForQuery
 */

func TestSelectConstant(t *testing.T) {
	// GIVEN: Active connection
	ctx := context.Background()
	conn, err := pgx.Connect(ctx, getConnectionConfig())
	require.NoError(t, err)
	defer conn.Close(ctx)

	// WHEN: Executing SELECT 1
	var result int
	err = conn.QueryRow(ctx, "SELECT 1").Scan(&result)

	// THEN: Should return 1
	require.NoError(t, err)
	assert.Equal(t, 1, result)
}

func TestMultiColumnSelect(t *testing.T) {
	// GIVEN: Active connection
	ctx := context.Background()
	conn, err := pgx.Connect(ctx, getConnectionConfig())
	require.NoError(t, err)
	defer conn.Close(ctx)

	// WHEN: Executing multi-column SELECT
	var num int
	var text string
	var floatVal float64

	err = conn.QueryRow(ctx, "SELECT 1 AS num, 'hello' AS text, 3.14 AS float_val").Scan(&num, &text, &floatVal)

	// THEN: Should return all columns correctly
	require.NoError(t, err)
	assert.Equal(t, 1, num)
	assert.Equal(t, "hello", text)
	assert.InDelta(t, 3.14, floatVal, 0.01)
}

func TestNullValues(t *testing.T) {
	// GIVEN: Active connection
	ctx := context.Background()
	conn, err := pgx.Connect(ctx, getConnectionConfig())
	require.NoError(t, err)
	defer conn.Close(ctx)

	// WHEN: Executing SELECT with NULL
	var nullCol *int
	var numCol int

	err = conn.QueryRow(ctx, "SELECT NULL AS null_col, 42 AS num_col").Scan(&nullCol, &numCol)

	// THEN: NULL should be handled correctly
	require.NoError(t, err)
	assert.Nil(t, nullCol)
	assert.Equal(t, 42, numCol)
}

func TestMultipleQueriesSequentially(t *testing.T) {
	// GIVEN: Active connection
	ctx := context.Background()
	conn, err := pgx.Connect(ctx, getConnectionConfig())
	require.NoError(t, err)
	defer conn.Close(ctx)

	// WHEN: Executing multiple queries
	var result1 int
	err = conn.QueryRow(ctx, "SELECT 1").Scan(&result1)
	require.NoError(t, err)

	var result2 string
	err = conn.QueryRow(ctx, "SELECT 'second query'").Scan(&result2)
	require.NoError(t, err)

	// THEN: Both queries should succeed
	assert.Equal(t, 1, result1)
	assert.Equal(t, "second query", result2)
}

func TestEmptyResultSet(t *testing.T) {
	// GIVEN: Active connection and empty table
	ctx := context.Background()
	conn, err := pgx.Connect(ctx, getConnectionConfig())
	require.NoError(t, err)
	defer conn.Close(ctx)

	// Create and empty table
	_, err = conn.Exec(ctx, "CREATE TABLE IF NOT EXISTS test_empty (id INT)")
	require.NoError(t, err)
	_, err = conn.Exec(ctx, "DELETE FROM test_empty")
	require.NoError(t, err)

	// WHEN: Querying empty table
	rows, err := conn.Query(ctx, "SELECT * FROM test_empty")
	require.NoError(t, err)
	defer rows.Close()

	// THEN: Result set should be empty
	assert.False(t, rows.Next(), "should have no rows")

	// Cleanup
	_, _ = conn.Exec(ctx, "DROP TABLE IF EXISTS test_empty")
}

func TestResultMetadata(t *testing.T) {
	// GIVEN: Active connection
	ctx := context.Background()
	conn, err := pgx.Connect(ctx, getConnectionConfig())
	require.NoError(t, err)
	defer conn.Close(ctx)

	// WHEN: Executing query
	rows, err := conn.Query(ctx, "SELECT 1 AS id, 'test' AS name")
	require.NoError(t, err)
	defer rows.Close()

	// THEN: Metadata should be available
	fieldDescriptions := rows.FieldDescriptions()
	require.Len(t, fieldDescriptions, 2)
	assert.Equal(t, "id", string(fieldDescriptions[0].Name))
	assert.Equal(t, "name", string(fieldDescriptions[1].Name))

	// Verify we have exactly one row
	assert.True(t, rows.Next())
	assert.False(t, rows.Next())
}

func TestStringWithSpecialCharacters(t *testing.T) {
	// GIVEN: Active connection and string with special characters
	ctx := context.Background()
	conn, err := pgx.Connect(ctx, getConnectionConfig())
	require.NoError(t, err)
	defer conn.Close(ctx)

	testString := `O'Reilly's "Book"`

	// WHEN: Querying with parameterized query
	var result string
	err = conn.QueryRow(ctx, "SELECT $1::text AS text", testString).Scan(&result)

	// THEN: String should be returned correctly
	require.NoError(t, err)
	assert.Equal(t, testString, result)
}

func TestParameterizedQueries(t *testing.T) {
	// GIVEN: Active connection
	ctx := context.Background()
	conn, err := pgx.Connect(ctx, getConnectionConfig())
	require.NoError(t, err)
	defer conn.Close(ctx)

	// WHEN: Executing parameterized query
	var num int
	var text string
	err = conn.QueryRow(ctx, "SELECT $1::int AS num, $2::text AS text", 42, "hello").Scan(&num, &text)

	// THEN: Parameters should be bound correctly
	require.NoError(t, err)
	assert.Equal(t, 42, num)
	assert.Equal(t, "hello", text)
}

func TestArrayResult(t *testing.T) {
	// GIVEN: Active connection
	ctx := context.Background()
	conn, err := pgx.Connect(ctx, getConnectionConfig())
	require.NoError(t, err)
	defer conn.Close(ctx)

	// WHEN: Executing query with multiple rows
	rows, err := conn.Query(ctx, `
		SELECT 1 AS id, 'first' AS name
		UNION ALL
		SELECT 2, 'second'
		UNION ALL
		SELECT 3, 'third'
	`)
	require.NoError(t, err)
	defer rows.Close()

	// THEN: Should return all rows
	var results []struct {
		ID   int
		Name string
	}

	for rows.Next() {
		var id int
		var name string
		err := rows.Scan(&id, &name)
		require.NoError(t, err)
		results = append(results, struct {
			ID   int
			Name string
		}{id, name})
	}

	require.Len(t, results, 3)
	assert.Equal(t, 1, results[0].ID)
	assert.Equal(t, 2, results[1].ID)
	assert.Equal(t, 3, results[2].ID)
}

func TestTransactionCommit(t *testing.T) {
	// GIVEN: Active connection and test table
	ctx := context.Background()
	conn, err := pgx.Connect(ctx, getConnectionConfig())
	require.NoError(t, err)
	defer conn.Close(ctx)

	_, err = conn.Exec(ctx, "CREATE TABLE IF NOT EXISTS test_commit (id INT, value VARCHAR(50))")
	require.NoError(t, err)
	defer conn.Exec(ctx, "DROP TABLE IF EXISTS test_commit")

	_, err = conn.Exec(ctx, "DELETE FROM test_commit")
	require.NoError(t, err)

	// WHEN: Transaction with commit
	tx, err := conn.Begin(ctx)
	require.NoError(t, err)

	_, err = tx.Exec(ctx, "INSERT INTO test_commit VALUES (1, 'committed')")
	require.NoError(t, err)

	err = tx.Commit(ctx)
	require.NoError(t, err)

	// THEN: Data should persist
	var count int
	err = conn.QueryRow(ctx, "SELECT COUNT(*) FROM test_commit").Scan(&count)
	require.NoError(t, err)
	assert.Equal(t, 1, count)
}

func TestTransactionRollback(t *testing.T) {
	// GIVEN: Active connection and test table
	ctx := context.Background()
	conn, err := pgx.Connect(ctx, getConnectionConfig())
	require.NoError(t, err)
	defer conn.Close(ctx)

	_, err = conn.Exec(ctx, "CREATE TABLE IF NOT EXISTS test_rollback (id INT, value VARCHAR(50))")
	require.NoError(t, err)
	defer conn.Exec(ctx, "DROP TABLE IF EXISTS test_rollback")

	_, err = conn.Exec(ctx, "DELETE FROM test_rollback")
	require.NoError(t, err)

	// WHEN: Transaction with rollback
	tx, err := conn.Begin(ctx)
	require.NoError(t, err)

	_, err = tx.Exec(ctx, "INSERT INTO test_rollback VALUES (1, 'will rollback')")
	require.NoError(t, err)

	err = tx.Rollback(ctx)
	require.NoError(t, err)

	// THEN: Data should NOT persist
	var count int
	err = conn.QueryRow(ctx, "SELECT COUNT(*) FROM test_rollback").Scan(&count)
	require.NoError(t, err)
	assert.Equal(t, 0, count)
}

func TestBatchQueries(t *testing.T) {
	// GIVEN: Active connection
	ctx := context.Background()
	conn, err := pgx.Connect(ctx, getConnectionConfig())
	require.NoError(t, err)
	defer conn.Close(ctx)

	// WHEN: Executing batch queries
	batch := &pgx.Batch{}
	batch.Queue("SELECT 1")
	batch.Queue("SELECT 2")
	batch.Queue("SELECT 3")

	br := conn.SendBatch(ctx, batch)
	defer br.Close()

	// THEN: All queries should execute
	var result1, result2, result3 int

	err = br.QueryRow().Scan(&result1)
	require.NoError(t, err)
	assert.Equal(t, 1, result1)

	err = br.QueryRow().Scan(&result2)
	require.NoError(t, err)
	assert.Equal(t, 2, result2)

	err = br.QueryRow().Scan(&result3)
	require.NoError(t, err)
	assert.Equal(t, 3, result3)
}

File: tests/client_compatibility/nodejs/query.test.js
import pg from 'pg';
const { Client } = pg;

/**
 * Test simple query execution via node-postgres driver.
 *
 * Tests P1 Simple Query Protocol:
 * - Query message handling
 * - IRIS SQL execution
 * - Row data encoding
 * - CommandComplete and ReadyForQuery
 */

const getConnectionConfig = () => ({
  host: process.env.PGWIRE_HOST || 'localhost',
  port: parseInt(process.env.PGWIRE_PORT || '5432'),
  database: process.env.PGWIRE_DATABASE || 'USER',
  user: process.env.PGWIRE_USERNAME || 'test_user',
  password: process.env.PGWIRE_PASSWORD || 'test',
  ssl: false,
});

describe('Simple Query Tests', () => {
  let client;

  beforeEach(async () => {
    client = new Client(getConnectionConfig());
    await client.connect();
  });

  afterEach(async () => {
    await client.end();
  });

  test('should execute SELECT constant', async () => {
    // WHEN: Executing SELECT 1
    const result = await client.query('SELECT 1');

    // THEN: Should return 1
    expect(result.rows).toHaveLength(1);
    expect(result.rows[0]['?column?']).toBe(1);
  });

  test('should execute multi-column SELECT', async () => {
    // WHEN: Executing multi-column SELECT
    const result = await client.query("SELECT 1 AS num, 'hello' AS text, 3.14 AS float_val");

    // THEN: Should return all columns correctly
    expect(result.rows).toHaveLength(1);
    expect(result.rows[0].num).toBe(1);
    expect(result.rows[0].text).toBe('hello');
    expect(result.rows[0].float_val).toBeCloseTo(3.14, 2);
  });

  test('should handle NULL values', async () => {
    // WHEN: Executing SELECT with NULL
    const result = await client.query('SELECT NULL AS null_col, 42 AS num_col');

    // THEN: NULL should be handled correctly
    expect(result.rows).toHaveLength(1);
    expect(result.rows[0].null_col).toBeNull();
    expect(result.rows[0].num_col).toBe(42);
  });

  test('should execute multiple queries sequentially', async () => {
    // WHEN: Executing multiple queries
    const result1 = await client.query('SELECT 1');
    const result2 = await client.query("SELECT 'second query'");

    // THEN: Both queries should succeed
    expect(result1.rows[0]['?column?']).toBe(1);
    expect(result2.rows[0]['?column?']).toBe('second query');
  });

  test('should handle empty result set', async () => {
    // GIVEN: Empty table
    await client.query('CREATE TABLE IF NOT EXISTS test_empty (id INT)');
    await client.query('DELETE FROM test_empty');

    try {
      // WHEN: Querying empty table
      const result = await client.query('SELECT * FROM test_empty');

      // THEN: Result set should be empty
      expect(result.rows).toHaveLength(0);
    } finally {
      // Cleanup
      await client.query('DROP TABLE IF EXISTS test_empty');
    }
  });

  test('should provide result metadata', async () => {
    // WHEN: Executing query
    const result = await client.query('SELECT 1 AS id, \'test\' AS name');

    // THEN: Metadata should be available
    expect(result.fields).toHaveLength(2);
    expect(result.fields[0].name).toBe('id');
    expect(result.fields[1].name).toBe('name');
    expect(result.rowCount).toBe(1);
  });

  test('should handle string with special characters', async () => {
    // GIVEN: String with special characters
    const testString = "O'Reilly's \"Book\"";

    // WHEN: Querying with parameterized query
    const result = await client.query('SELECT $1::text AS text', [testString]);

    // THEN: String should be returned correctly
    expect(result.rows[0].text).toBe(testString);
  });

  test('should support parameterized queries', async () => {
    // WHEN: Executing parameterized query
    const result = await client.query('SELECT $1::int AS num, $2::text AS text', [42, 'hello']);

    // THEN: Parameters should be bound correctly
    expect(result.rows[0].num).toBe(42);
    expect(result.rows[0].text).toBe('hello');
  });

  test('should handle array result', async () => {
    // WHEN: Executing query with multiple rows
    const result = await client.query(`
      SELECT 1 AS id, 'first' AS name
      UNION ALL
      SELECT 2, 'second'
      UNION ALL
      SELECT 3, 'third'
    `);

    console.log('UNION result columns:', result.fields.map(f => f.name));
    console.log('UNION result row:', JSON.stringify(result.rows[0], null, 2));

    // THEN: Should return all rows
    expect(result.rows).toHaveLength(3);
    expect(result.rows[0].id).toBe(1);
    expect(result.rows[1].id).toBe(2);
    expect(result.rows[2].id).toBe(3);
  });
});

describe('Transaction Tests', () => {
  let client;

  beforeEach(async () => {
    client = new Client(getConnectionConfig());
    await client.connect();
  });

  afterEach(async () => {
    await client.end();
  });

  test('should support basic commit', async () => {
    // GIVEN: Test table
    await client.query('CREATE TABLE IF NOT EXISTS test_commit (id INT, value VARCHAR(50))');

    try {
      await client.query('DELETE FROM test_commit');

      // WHEN: Transaction with commit
      await client.query('BEGIN');
      await client.query("INSERT INTO test_commit VALUES (1, 'committed')");
      await client.query('COMMIT');

      // THEN: Data should persist
      const result = await client.query('SELECT COUNT(*) FROM test_commit');
      expect(parseInt(result.rows[0].count)).toBe(1);
    } finally {
      // Cleanup
      await client.query('DROP TABLE IF EXISTS test_commit');
    }
  });

  test('should support basic rollback', async () => {
    // GIVEN: Test table
    await client.query('CREATE TABLE IF NOT EXISTS test_rollback (id INT, value VARCHAR(50))');

    try {
      await client.query('DELETE FROM test_rollback');

      // WHEN: Transaction with rollback
      await client.query('BEGIN');
      await client.query("INSERT INTO test_rollback VALUES (1, 'will rollback')");
      await client.query('ROLLBACK');

      // THEN: Data should NOT persist
      const result = await client.query('SELECT COUNT(*) FROM test_rollback');
      expect(parseInt(result.rows[0].count)).toBe(0);
    } finally {
      // Cleanup
      await client.query('DROP TABLE IF EXISTS test_rollback');
    }
  });
});
